Multi node support

flaeppe opened the pull request:
## Backwards compatibility

The changes here are not backwards compatible in the (odd?) case of previously having had this playbook run on multiple hosts.
We'd now expect a hosts file to refer to any nodes involved in a _single_ swarm.

Backwards compatibility can be achieved, when above is false, by declaring any hosts deployed with the previous version under both `docker_swarm_managers` and `docker_swarm_workers`:

e.g.

Previous hosts file:
```
[managers]
host1
```

New hosts file:

```
[managers]
host1

[docker_swarm_managers]
host1

[docker_swarm_workers]
host1
```

As far as I see, it should only be necessary to change the hosts file.

## Additional info

- The swarm.yml task set asserts that there's an _odd_ number of managers declared in the `docker_swarm_managers` group.

- Node labels can be declared via the `docker_swarm_labels` (per host/group/etc) if wanted

- Three new (default) variables: `docker_swarm_interface`, `docker_swarm_addr` and `docker_swarm_port` have been added

- Would it be overkill to add a guard that validates that the worker group is non empty if all managers are declared as "manager only" nodes?

## Some drawbacks

### Demoting a manager node

As far as I could dig, I found no reasonable task/steps to let the playbook demote a manager (i.e. removing host from the manager group while having it in the worker group). Although handling that here could be quite an edge case, or perhaps at most a nice to have.

As of now, one could rearrange the hosts file so that the host only appears in the worker group and then _manually_ run `docker node demote NODE` on the manager node to sync the swarm with the hosts file layout.

### hosts groups declaration order

~As some steps are run with flags `run_once` and `delegate_to` we expect the `docker_swarm_managers` group to be declared before the `docker_swarm_workers` group. Otherwise I suspect that a `register` variable is undefined and stuff crashes.~

~At least that's what I could find online. I haven't actually tested with a real hosts file just yet.~

~**We should really test if this assumption is correct, would be nice if it isn't.**~

**EDIT: This should no longer be an issue. See changes in: a9164fd for how it was resolved. Worth noting is also that the playbook with those fixes has been tested with a multi node setup (1 worker node, 1 manager only node) where the worker group appeared before the manager group in the hosts file**

flaeppe committed:
Add __pycache__ to .gitignore

flaeppe committed:
Implement tests for configuring multiple workers/managers

flaeppe committed:
Extend molecule.yml platforms

- Declare one manager node
- Declare one worker (only) node
- Create a new second manager node

flaeppe committed:
Implement tasks for joining worker/manager nodes via hosts groups

flaeppe committed:
Implement swarm node labelling task

- Handle an optional host variable 'docker_swarm_labels' (dict)

flaeppe committed:
Implement task controlling manager availability

- Requirement to explicitly declare a node in both manager and worker
  groups to set a node as both worker and manager
- Allows us to create manager only nodes
- Per default a host only in the manager group is not a worker node

flaeppe committed:
Require an odd number of manager nodes to be declared

- Guards against declaring 'docker_swarm_managers' group as any even count of hosts

flaeppe committed:
Update README

flaeppe committed:
Correct untrue comment(s)

kjagiello commented on tasks/swarm.yml:
```diff
@@ -1,9 +1,55 @@
+    advertise_addr: "{{ docker_swarm_addr }}"
+  # 'run_once' and 'delegate_to' allows us to register 'swarm' for all hosts
+  # Although this relies on declaration order of the hosts in order to work
+  # i.e. swarm manager group has to come before the worker group
```
We should verify that it actually is true.

flaeppe committed:
Add comment to help trace where an expected assert comes from

flaeppe committed:
Set better `hosts` in README:s example playbook(s)

flaeppe committed:
Utilise facts to remove dependency on group appearance in hosts file

- 'docker_swarm_workers' and 'docker_swarm_managers' can now appear in
  any order in a hosts file

flaeppe committed:
fixup! Update README

flaeppe commented:
I have now tested the playbook with a multi node setup(1 worker node, 1 manager only node) successfully.

(`hosts groups declaration order` is no longer an issue either, this was also tested)

flaeppe committed:
Use same certificates on all manager nodes

flaeppe committed:
Delegate setting manager availability to master node

flaeppe committed:
Replace molecule 'debug' flag with 'travis_wait'

- Idempotence step has no debug output anyway

flaeppe committed:
Give test run 30 min (with no log output)

- Default 'travis_wait' increase at 20 min seems to be just on the edge

beshrkayali commented:
Nitpick: Would be nice to make tags either use dash `-` or `_` as a word separator. I think all existing ones use a `-` so maybe we should make swarm task tags also use that? :)
